### PR TITLE
Added gears.string.psplit function to support patterns

### PR DESCRIPTION
**Pull request by Piterden**
It will support strings with more than one char length and regex patterns. Also there's no need for additional checks for start and end of a string anymore.

### CODE REVIEW

**Comment by actionless** [comment]
you need to add new unit tests for newly-added functionality

**Comment by actionless** [comment]
(in spec/gears/string_spec.lua)

File: spec/gears/string_spec.lua [hunk 1]
```diff
@@ -51,11 +51,17 @@ describe("gears.string", function()
...
+        assert.is_same(gstring.split("foo.", "%."), {"foo", ""})
+        assert.is_same(gstring.split("foo.bar", "%."), {"foo", "bar"})
+
         assert.is_same(gstring.split("", "."), {""})
-        assert.is_same(gstring.split(".", "."), {"", ""})
-        assert.is_same(gstring.split("foo", "."), {"foo"})
-        assert.is_same(gstring.split("foo.", "."), {"foo", ""})
-        assert.is_same(gstring.split("foo.bar", "."), {"foo", "bar"})
+        assert.is_same(gstring.split("a", "."), {"", ""})
+        assert.is_same(gstring.split("foo", "."), {"", "", "", ""})
```
Comment (actionless) [inline spec/gears/string_spec.lua:62]: this is not normal - API compatibility shouldn't be broken within major release

**Review by actionless — CHANGES_REQUESTED** [review]

File: spec/gears/string_spec.lua (hunk 1 above)
Comment (Piterden) [reply spec/gears/string_spec.lua:62]: Okay, what about to add third parameter? The flag which will enable pattern matching instead of string.

Comment (Piterden) [reply spec/gears/string_spec.lua:62]: Or would it be better to create a separate function like `psplit`?

Comment (actionless) [reply spec/gears/string_spec.lua:62]: depending on implementation - as far as i see from the diff implementation almost totally different, so i think 2 separate functions would be alright

Comment (actionless) [reply spec/gears/string_spec.lua:62]: also as far as i see from codecov warnings - not all the code branches are covered by the new tests

Comment (Piterden) [reply spec/gears/string_spec.lua]: Done.

**Review by actionless — APPROVED** [review]

**Comment by Piterden** [comment]
@actionless there are two workflows haven't been passed. Should I do something with them to fix?
image

**Review by Aire-One — APPROVED** [review]

**Comment by Aire-One** [comment]
> @actionless there are two workflows haven't been passed. Should I do something with them to fix? image

The luajit error was from a flaky test. I restarted the CI and it's green now.

The fixed-lgi-lua5.2 is a luacheck warning:

```
Checking lib/awful/permissions/init.lua           1 warning

    lib/awful/permissions/init.lua:371:30: Error prone negation: negation is executed before relational operator.
```

That's a new warning they added recently, https://github.com/lunarmodules/luacheck/pull/94.

It needs to be solved first.

**Comment by Aire-One** [comment]
@Piterden, with https://github.com/awesomeWM/awesome/pull/3840 now merged, can you please rebase/merge master into your branch?

**Comment by Piterden** [comment]
@Aire-One excuse me for a long delay. Done!

**Review by Aire-One — APPROVED** [review]

**Review by actionless — APPROVED** [review]

**Comment by Piterden** [comment]
What should I do to fix checks passing?

**Comment by actionless** [comment]
@Piterden https://github.com/awesomeWM/awesome/actions/runs/6820112574/job/18557186346?pr=3839#step:34:835

**Review by Elv13 — APPROVED** [review]

**Comment by Elv13** [comment]
I squashed the commits, this should solve the problem which was preventing this from auto-merging:

```
Error -> lib/gears/filesystem.lua @ 12
suite spec/awful/completion_spec.lua
lib/gears/filesystem.lua:12: error loading module 'gears.string' from file 'lib/gears/string.lua':
	lib/gears/string.lua:101: unexpected symbol near 'for'

stack traceback:
	lib/gears/filesystem.lua:12: in main chunk


Error -> spec/awful/prompt_spec.lua @ 63
main
lib/gears/filesystem.lua:12: error loading module 'gears.string' from file 'lib/gears/string.lua':
	lib/gears/string.lua:101: unexpected symbol near 'for'
```

**Pull request merged**